src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs

```diff
@@ -169,173 +154,57 @@ private unsafe static OpenedReader TryOpenReaderForInMemoryPdb(IntPtr inMemoryPd
                 return null;
             }
 
-            OpenedReader result = null;
-            MetadataReaderProvider provider = null;
+            var provider = MetadataReaderProvider.FromMetadataImage((byte*)inMemoryPdbAddress, inMemoryPdbSize);
             try
             {
-                provider = MetadataReaderProvider.FromMetadataImage((byte*)inMemoryPdbAddress, inMemoryPdbSize);
-                result = new OpenedReader(provider, provider.GetMetadataReader());
+                // may throw if the metadata is invalid
+                provider.GetMetadataReader();
+                return provider;
             }
             catch (BadImageFormatException)
             {
+                provider.Dispose();
                 return null;
             }
-            finally
-            {
-                if (result == null)
-                {
-                    provider?.Dispose();
-                }
-            }
-
-            return result;
         }
 
-        private static OpenedReader TryOpenReaderFromAssemblyFile(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize)
+        private unsafe static PEReader TryGetPEReader(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize)
         {
-            Debug.Assert(assemblyPath != null);
-            Debug.Assert(loadedPeAddress != IntPtr.Zero);
-
-            // TODO: When/if PEReader supports reading loaded PE files use loadedPeAddress and loadedPeSize instead of opening the file again.
-            Stream peStream = TryOpenFile(assemblyPath);
-            if (peStream == null)
+            if (loadedPeAddress != IntPtr.Zero && loadedPeSize > 0)
             {
-                return null;
+                return new PEReader((byte*)loadedPeAddress, loadedPeSize, isLoadedImage: true);
```

Comment:
I see. Yes, this won't throw. The constructor doesn't read the memory, It just creates internal structures that hold onto it. The actual reading later is guarded against bad image.